ring.go

```diff
@@ -39,7 +39,8 @@ type ringBackend interface {
 	// start returns the index of first element that can be read.
 	start() uint
 
-	// end returns the index of the last element that can be read.
+	// end returns the index after the last element that can be
+	// read. It is compatible with go [:] slice expressions.
```

Comment:
I think this comment is what caused alllllll my confusion, leading up to the off-by-one. Hopefully fixing the internal docs here will prevent more of that (: